core/arch/x86/encode.c

```diff
@@ -1044,6 +1064,9 @@ reg_size_ok(decode_info_t *di /*prefixes field is IN/OUT; x86_mode is IN*/, reg_
                 opsize == OPSZ_1 || opsize == OPSZ_2 || opsize == OPSZ_4 ||
                 opsize == OPSZ_8);
     }
+    if (optype == TYPE_T_REG || optype == TYPE_T_MODRM) {
```

Comment:
Two issues here:

1) I think this code is not correct and neither is the TYPE_K* code above: it needs to look at `reg`.  As is, it only does a self-sanity check on our own decode table entries, and ignores the instruction being encoded.  Please look at the TYPE_K* code as well.

2) I feel like this shouldn't be needed here: shouldn't the simple MxN have already been resolved?  If resolve_var_x64_size() is not already called on this path maybe it should be?  How do the existing OPSZ_4x8 entries work?

Reply:
Reverse order: 

2. with resolve_var_x64_size() fixed, the code above was removed.

1. I have added the following comment:

```
   /* XXX i#1312: we haven't really resolved the register's size. The mask register                                                                                                                               
     * size is dependent on the machine's support for certain AVX-512 derivatives. We may                                                                                                                          
     * want to look at XCR0 bits and resolve the size in a new function similar to                                                                                                                                 
     * resolve_var_x64_size(). At the same time, we would want to change the mask register                                                                                                                         
     * sizes in the decode table reflecting a simple variable size for all instructions.                                                                                                                           
     */
```